[JIT] Pass To Safely Remove Aten Inplace Ops

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33297 [JIT] Allow mutated values as functional graph inputs
* #33199 [JIT] remove list appends
* **#33186 [JIT] Pass To Safely Remove Aten Inplace Ops**
* #33020 [JIT] Functional Graph Pass

This helps create larger functional graphs. It has the potential to increase memory use, so in order to land this on by default we would probably also do a reuse of buffers pass.

This is currently O(n * | Removed Nodes | ) because we have to rebuild the alias Db each time we make a change. This pass is critical to creating functional graphs, so this might be a compelling use case to build incremental updates to alias Db.

Differential Revision: [D20603189](https://our.internmc.facebook.com/intern/diff/D20603189)

Commit:
add pass to remove aten mutation ops when safe

[ghstack-poisoned]

Commit:
Update on "[JT] [WIP] Pass To Safely Remove Mutation"




[ghstack-poisoned]

Commit:
Update on "[JIT] Pass To Safely Remove Aten Inplace Ops"


This helps create larger functional graphs. It has the potential to increase memory use, so in order to land this on by default we would probably also do a reuse of buffers pass.



[ghstack-poisoned]

Commit:
Update on "[JIT] Pass To Safely Remove Aten Inplace Ops"


This helps create larger functional graphs. It has the potential to increase memory use, so in order to land this on by default we would probably also do a reuse of buffers pass.



[ghstack-poisoned]

Commit:
Update on "[JIT] Pass To Safely Remove Aten Inplace Ops"


This helps create larger functional graphs. It has the potential to increase memory use, so in order to land this on by default we would probably also do a reuse of buffers pass.

This is currently O(n * | Removed Nodes | ) because we have to rebuild the alias Db each time we make a change. This pass is critical to creating functional graphs, so this might be a compelling use case to build incremental updates to alias Db.

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/create_functional_graphs.cpp:
```diff
@@ -212,6 +212,97 @@ void InlineFunctionalGraphs(Block* block) {
+      node->destroy();
+
+      // :(  TODO: incremental update ?
+      aliasDb_ = torch::make_unique<AliasDb>(graph_);
```
I'm working on this part but the patch can still be reviewed

suo commented on torch/csrc/jit/passes/create_functional_graphs.cpp:
```diff
@@ -212,6 +212,97 @@ void InlineFunctionalGraphs(Block* block) {
+
+ private:
+  bool uniqueAlias(Value* v) {
+    // if the output isn't contained or alias by the inputs to its node, it's
```
this references an "output", but seems like `v` actually is an input.

suo commented on torch/csrc/jit/passes/create_functional_graphs.cpp:
```diff
@@ -212,6 +212,97 @@ void InlineFunctionalGraphs(Block* block) {
+
+      auto schema_name = node->schema().name();
+      auto new_schema = schema_name.substr(0, schema_name.size() - 1);
+      auto new_node = graph_->create(Symbol::fromQualString(new_schema), 1);
```
Let's propagate the source/scope information as well.

suo commented on torch/csrc/jit/passes/create_functional_graphs.cpp:
```diff
@@ -212,6 +212,97 @@ void InlineFunctionalGraphs(Block* block) {
+
+      Value* mutated_value = node->inputs().at(0);
+
+      // We can only remove mutation to values that are unique aliases in the
```
I think the concept of `unique` introduces confusion—I would expect it to mean "this value is the only pointer to this tensor memory" (a la `unique_ptr`). But in this case, if I had:
```
z = add(x, y)
a = z[0]
add_(z, …)
```
It appears that `uniqueAlias(z)` would return true, even though `a` points to `z`. 

It seems the rule we want is: 
1. We witnessed the creation of `foo`'s memory and know that it isn't an alias of anything. In implementation terms, `foo` is a `MemoryLocation` (a leaf on the points-to graph).
2. We can "fold" the in-place add into the creation of `foo`. In implementation terms, we can move `add_` after `foo->node()` in a topologically valid way.

Can we modify this pass to check `MemoryLocation` instead of using `uniqueAlias`?

Also: it would be helpful to have a high-level comment describing the rules. The comments here are good but somewhat in the weeds (they describe what we *can't* do in the name of safety, but don't present the overall model of what we *want* to do, if that makes any sense…)

eellison replied:
I don't think the case you've added here would actually induces a changes in behavior. This would fail because the creation and subsequent mutation could not be made atomic. 

Sure, i can add more description & high-level commentary.

eellison replied:
I'll look into using the MemoryLocation here

eellison replied:
What you've described in 1 and 2 is the same as what already exists, using `MemoryLocations` instead of the aliasDb to check if an op is the creation of a tensor..

eellison replied:
Actually, I take that back. You've added the additional check that: `"this value is the only pointer to this tensor memory" (a la unique_ptr)`, however this is an unnecessary restriction.

suo replied:
Sorry I should be a little clearer: I think what I've suggested is the same functionally as what you have here, it just re-uses an existing concept (`MemoryLocation`) instead of adding a new one ("uniqueness"). So really it's just shuffling around concepts to make the code a little clearer (to me at least).

When I was talking about "this value is the only tensor memory"—I'm not saying that should be the requirement, I'm just saying that that's what I assume "unique" means given what it means in other contexts, which is confusing because that's not precisely what we want here, as you've pointed out.

suo replied:
So concretely, the change I am suggesting is: rewrite using an `isMemoryLocation` function instead of `uniqueAlias`.

suo commented on torch/csrc/jit/passes/create_functional_graphs.h:
```diff
@@ -10,5 +10,7 @@ TORCH_API void CreateFunctionalGraphs(const std::shared_ptr<Graph>& graph);
 
 TORCH_API void InlineFunctionalGraphs(const std::shared_ptr<Graph>& graph);
 
+TORCH_API void RemoveMutation(const std::shared_ptr<Graph>& graph);
```
Can we get a comment describing at a high-level the purpose of this pass, and generally what is a candidate for mutation removal?

Commit:
Update on "[JIT] Pass To Safely Remove Aten Inplace Ops"


This helps create larger functional graphs. It has the potential to increase memory use, so in order to land this on by default we would probably also do a reuse of buffers pass.

This is currently O(n * | Removed Nodes | ) because we have to rebuild the alias Db each time we make a change. This pass is critical to creating functional graphs, so this might be a compelling use case to build incremental updates to alias Db.

suo commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -151,6 +151,24 @@ bool AliasDb::writesToAlias(Node* n, const ValueSet& vs) const {
   return false;
 }
 
+bool AliasDb::writtenToAtNode(const at::ArrayRef<Value*> vs, Node* n) const {
```
This seems like it's already implemented in `writesToAlias`.

Commit:
Update on "[JIT] Pass To Safely Remove Aten Inplace Ops"


This helps create larger functional graphs. It has the potential to increase memory use, so in order to land this on by default we would probably also do a reuse of buffers pass.

This is currently O(n * | Removed Nodes | ) because we have to rebuild the alias Db each time we make a change. This pass is critical to creating functional graphs, so this might be a compelling use case to build incremental updates to alias Db.

Commit:
Update on "[JIT] Pass To Safely Remove Aten Inplace Ops"


This helps create larger functional graphs. It has the potential to increase memory use, so in order to land this on by default we would probably also do a reuse of buffers pass.

This is currently O(n * | Removed Nodes | ) because we have to rebuild the alias Db each time we make a change. This pass is critical to creating functional graphs, so this might be a compelling use case to build incremental updates to alias Db.

Commit:
Update on "[JIT] Pass To Safely Remove Aten Inplace Ops"


This helps create larger functional graphs. It has the potential to increase memory use, so in order to land this on by default we would probably also do a reuse of buffers pass.

This is currently O(n * | Removed Nodes | ) because we have to rebuild the alias Db each time we make a change. This pass is critical to creating functional graphs, so this might be a compelling use case to build incremental updates to alias Db.

Differential Revision: [D20603189](https://our.internmc.facebook.com/intern/diff/D20603189)
